uint8 id0
float32 angle0
uint8 id1
float32 angle1